tests/posix/pthread_pressure/src/main.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2023, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pthread.h>
+
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define STACK_SIZE K_THREAD_STACK_LEN(CONFIG_TEST_STACK_SIZE)
+
+/* update interval for printing stats */
+#if CONFIG_TEST_DURATION_S >= 60
+#define UPDATE_INTERVAL_S 10
+#elif CONFIG_TEST_DURATION_S >= 30
+#define UPDATE_INTERVAL_S 5
+#else
+#define UPDATE_INTERVAL_S 1
+#endif
+
+/* 32 threads is mainly a limitation of find_lsb_set() */
+#define NUM_THREADS MIN(32, MIN(CONFIG_TEST_NUM_CPUS, CONFIG_MAX_PTHREAD_COUNT))
+
+typedef int (*create_fn)(int i);
+typedef int (*join_fn)(int i);
+
+static void *setup(void);
+static void before(void *fixture);
+
+/* bitmask of available threads */
+static bool alive[NUM_THREADS];
+
+/* array of thread stacks */
+static K_THREAD_STACK_ARRAY_DEFINE(thread_stacks, NUM_THREADS, STACK_SIZE);
+
+static struct k_thread k_threads[NUM_THREADS];
+static size_t counters[NUM_THREADS];
+
+static void print_stats(uint64_t now, uint64_t end)
+{
+	printk("now (ms): %llu end (ms): %llu\n", now, end);
+	for (int i = 0; i < NUM_THREADS; ++i) {
+		printk("Thread %d created and joined %zu times\n", i, counters[i]);
+	}
+}
+
+static void test_create_join_common(const char *tag, create_fn create, join_fn join)
+{
+	int i;
+	__unused int ret;
+	uint64_t now_ms = k_uptime_get();
+	const uint64_t end_ms = now_ms + MSEC_PER_SEC * CONFIG_TEST_DURATION_S;
+	uint64_t update_ms = now_ms + MSEC_PER_SEC * UPDATE_INTERVAL_S;
+
+	printk("BOARD: %s\n", CONFIG_BOARD);
+	printk("NUM_THREADS: %u\n", NUM_THREADS);
+	printk("TEST_NUM_CPUS: %u\n", CONFIG_TEST_NUM_CPUS);
+	printk("TEST_DURATION_S: %u\n", CONFIG_TEST_DURATION_S);
+	printk("TEST_DELAY_US: %u\n", CONFIG_TEST_DELAY_US);
+
+	for (i = 0; i < NUM_THREADS; ++i) {
+		/* spawn thread i */
+		ret = create(i);
+#ifdef CONFIG_TEST_EXTRA_ASSERTIONS
+		zassert_ok(ret, "%s_create(%d)[%zu] failed: %d", tag, i, counters[i], ret);
+#endif
```

Comment:
Instead of adding `__unused` attribute to `int ret`, I would suggest using `if (IS_ENABLED(...))` instead of `#ifdef` so that the compiler is aware of the execution path that uses `ret`.

Reply:
Yea... I was really on the fence there. In theory, the linker should discard any dead code, and perhaps it's just me being paranoid, but I found that the #ifdef seemed to result in fewer failed assertions.

I'll respin. Ultimately though, I want @jgl-meta to make the call whether this fix is merged first or whether he will hold out for pthreads to be water tight.

Ideally we could do the latter.

Reply:
> Yea... I was really on the fence there. In theory, the linker should discard any dead code, and perhaps it's just me being paranoid, but I found that the #ifdef seemed to result in fewer failed assertions.

Yes, compiler should optimise out any dead code. [Linux kernel coding style also recommends using `IS_ENABLED` instead of preprocessor conditionals whenever possible](https://www.kernel.org/doc/html/latest/process/coding-style.html#conditional-compilation) for this reason -- maybe it is about the time we document this in the Zephyr coding guidelines ...

Reply:
> I'll respin. Ultimately though, I want @jgl-meta to make the call whether this workaround / test is merged first or whether he will hold out for pthreads to be water tight

I think this can go ahead and be merged. I think it is better to have this than make no mention of it at all.

Reply:
Made @stephanosio 's suggestions.